### PR TITLE
Add support for overriding color palette

### PR DESCRIPTION
**Pull request by mallport**
Add functionality to override elements of the existing color palette.

The environment variable `REACT_APP_PALETTE_OVERRIDE` is written as single line JSON, corresponding to the fields in the defaultPalette in [onyxia-ui/color.ts](https://github.com/InseeFrLab/onyxia-ui/blob/main/src/lib/color.ts). The contents then patch the values in the defaultPalette.

### CODE REVIEW

File: src/ui/theme.tsx [hunk 1]
```diff
@@ -89,15 +90,23 @@ const { ThemeProvider, useTheme } = createThemeProvider({
...
-                case "ultraviolet":
-                    return ultravioletPalette;
-                case "verdant":
-                    return verdantPalette;
+            const paletteOverrides = getPaletteOverride();
+            if (paletteOverrides) {
+                return {
+                    ...defaultPalette,
+                    ...paletteOverrides
+                };
```
Comment (garronej) [inline src/ui/theme.tsx]: In this instance, a deep merge would be the most suitable approach.

Comment (garronej) [reply src/ui/theme.tsx]: I believe the anticipated outcome is for the selected palette to be overwritten. For instance, an administrator may prefer the 'France' theme, yet wish to modify just a single specific color.

File: .env [hunk 2]
```diff
@@ -21,6 +21,10 @@ REACT_APP_TERMS_OF_SERVICES=
 
 # Can be one of onyxia, france or ultraviolet
 REACT_APP_THEME_ID=onyxia
+# Overrides fields of the color palette, needs to be single line JSON corresponding to the fields in paletteOverride
+# If this is set, then REACT_APP_THEME_ID is ignored
+# Example: '{"focus": {"main": "#00824D","light": "#00824D","light2": "#1A9D49"}}'
```
Comment (garronej) [inline .env]:
```suggestion
# This parameter allows you to override some, or all, of the color palette values for the selected theme.
#
# Please adhere to the JSON format outlined in the following object:
# https://github.com/InseeFrLab/onyxia-ui/blob/0fa8667227aac328cb685cc140f3cd2186956304/src/lib/color.ts#L15-L53
#
# Should you desire to personalize the color palette without the assistance of a UI designer, 
# there is a simple way. You can copy and paste the default palette and request ChatGPT to 
# propose a variation. For instance, you might want to emphasize a specific color, like 
# "TypeScript's blue". ChatGPT can use that as a reference to generate a new color scheme 
# tailored to your preference.
#
# Example:  
#
# PALETTE_OVERRIDE: |
#   { 
#     "focus": { 
#         "main": "#00824D", 
#         "light": "#00824D"
#      } 
#   }
```

The JSON must be formatted as a single line in the .env file, but we anticipate that users will define these variables from the values.json of their Helm charts. That's why we provide examples in this particular format.

**Review by garronej — CHANGES_REQUESTED** [review]
Thank you very much for your valuable contribution.

Please find my review along with some suggested modifications attached.

Looking forward to incorporating your changes!

File: src/ui/theme.tsx (hunk 1 above)
Comment (mallport) [reply src/ui/theme.tsx]: Thanks for your suggestions Joseph! The proposed changes are included in [55d8cc1](https://github.com/InseeFrLab/onyxia-web/pull/565/commits/55d8cc15bf343276491e9b6611f9d258ec3eb2a3) :)

File: src/ui/theme.tsx [hunk 3]
```diff
@@ -72,6 +72,8 @@ import { ComponentType } from "ui/tools/types/ComponentType";
 import type { Language } from "ui/i18n";
 import { createOnyxiaSplashScreenLogo } from "onyxia-ui/lib/SplashScreen";
 import { THEME_ID } from "keycloak-theme/login/envCarriedOverToKc";
+import { getPaletteOverride } from "./env";
+import _ from "lodash";
```
Comment (garronej) [inline src/ui/theme.tsx]: I'd rather avoid introducing lodash.  
Please create a util in `src/ui/tools`. It's where we put all the general utilities.

File: src/ui/theme.tsx [hunk 4]
```diff
@@ -201,6 +192,30 @@ export function createGetViewPortConfig(params: {
     return { getViewPortConfig };
 }
 
+function getPalette(themeId: typeof THEME_ID, paletteOverrides?: Object): PaletteBase {
```
Comment (garronej) [inline src/ui/theme.tsx]: Please use a cosure, we prefer avoiding creating function that are used only once.

File: src/ui/theme.tsx [hunk 5]
```diff
@@ -201,6 +192,30 @@ export function createGetViewPortConfig(params: {
...
+    var palette: PaletteBase;
```
Comment (garronej) [inline src/ui/theme.tsx]:
```suggestion
    let palette: PaletteBase;
```

File: src/ui/theme.tsx [hunk 6]
```diff
@@ -201,6 +192,30 @@ export function createGetViewPortConfig(params: {
...
+        case "ultraviolet":
+            palette = ultravioletPalette;
+            break;
+        case "verdant":
+            palette = verdantPalette;
+            break;
+    }
+
+    if (paletteOverrides) {
+        _.merge(palette, paletteOverrides);
```
Comment (garronej) [inline src/ui/theme.tsx]: This is destructive for the default palette.  
In practice it's no big deal but it's an undesired side effect.

**Review by garronej — CHANGES_REQUESTED** [review]
Hey @mmwinther,   
How are you? Thanks for the contribution!  
Here is my review.

**Comment by mmwinther** [comment]
@garronej thanks for the input! We've been having a great summer here in Norway.

See the latest changes which should address your comments 👍

**Comment by garronej** [comment]
Thanks a lot for your contribution!

**Comment by garronej** [comment]
The dragoon now changes color as well:  

https://github.com/InseeFrLab/onyxia-web/assets/6702424/18a11df1-986f-4722-95d8-ed2bf1fc60be

**Comment by garronej** [comment]
The palette override now applies to the favicon:  

image

And is carried over to the login pages:  

image

**Comment by garronej** [comment]
All this is available in onyxia-web 2.23

**Pull request merged**